### PR TITLE
Support user-supplied profiles

### PR DESCRIPTION
**Pull request by jlaneve**
## Description

This PR extends the `ProfileConfig` interface to support two modes:
1. the profile mapping, which takes an Airflow connection and translates it to a dbt profile
2. a user-supplied profile file

When using a profile mapping, you can do one of the following:

1. Use the AutomaticProfileMapping class to auto-select a profile mapping

```python
ProfileConfig(
    profile_name="default",
    target_name="dev",
    profile_mapping=get_automatic_profile_mapping(
        conn_id="airflow_db",
        args={"schema": "public"},
    ),
)
```

2. Specify a specific profile mapping

```python
ProfileConfig(
    profile_name="default",
    target_name="dev",
    profile_=PostgresUserPasswordProfileMapping(
        conn_id="airflow_db",
        profile_args={"schema": "public"},
    ),
)
```

When using your own profile, you can supply the `path_to_profiles_yml`:

```python
ProfileConfig(
    profile_name="default",
    target_name="dev",
    path_to_profiles_yml=DBT_ROOT_PATH / "jaffle_shop" / "profiles.yml",
)
```

## Related Issue(s)

closes https://github.com/astronomer/astronomer-cosmos/issues/336
closes https://github.com/astronomer/astronomer-cosmos/issues/269

## Breaking Change?

This changes the `ProfileConfig` class, which isn't yet user-facing.

## Checklist

- [ ] I have made corresponding changes to the documentation (if required)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by tatiana — COMMENTED** [review]
I've successfully run `basic_cosmos_dag` from this branch.

Some comments are in line. Given the timelines, I added "major" to the ones I think are more relevant.

File: dev/dags/example_virtualenv.py [hunk 1]
```diff
@@ -6,24 +6,28 @@
...
 
 from cosmos import DbtDag, ExecutionMode, ExecutionConfig, ProjectConfig, ProfileConfig
+from cosmos.profiles import PostgresUserPasswordProfileMapping
 
 DEFAULT_DBT_ROOT_PATH = Path(__file__).parent / "dbt"
 DBT_ROOT_PATH = Path(os.getenv("DBT_ROOT_PATH", DEFAULT_DBT_ROOT_PATH))
 PROJECT_NAME = "jaffle_shop"
-CONNECTION_ID = "airflow_db"
+
+profile_config = ProfileConfig(
```
[the remark this answers is not in the document]
Comment (harels) [reply dev/dags/example_virtualenv.py:15]: going to address this in a followup PR

File: tests/profiles/exasol/test_exasol_user_pass.py [hunk 2]
```diff
@@ -185,8 +191,9 @@ def test_dsn_formatting() -> None:
         schema="my_database",
     )
 
-    profile_mapping = ExasolUserPasswordProfileMapping(conn, {"schema": "my_schema", "threads": 1})
-    assert profile_mapping.dsn == "my_host:1000"
+    with patch("airflow.hooks.base.BaseHook.get_connection", return_value=conn):
```
[the remark this answers is not in the document]
Comment (harels) [reply tests/profiles/exasol/test_exasol_user_pass.py:194]: going to address this in a followup PR

**Review by harels — APPROVED** [review]

File: dev/dags/example_virtualenv.py [hunk 3]
```diff
@@ -6,24 +6,28 @@
...
+    profile_name="default",
+    target_name="dev",
+    profile_mapping=PostgresUserPasswordProfileMapping(
+        conn_id="airflow_db",
+        profile_args={"schema": "public"},
```
[the remark this answers is not in the document]
Comment (harels) [reply dev/dags/example_virtualenv.py:20]: going to address this in a followup PR

**Pull request merged**